Timestamp range in GetAccountTransfers filter (#1470)
tigerbeetle/tigerbeetle: docs/reference/operations/get_account_transfers.md

Reviewer: Hmm, what if both `timestamp_min` and `timestamp_max` were _exclusive_ instead of inclusive?
That way pagination still works as before.

...I think (more generally) I am confused whether it is good/safe for an application to use tigerbeetle's hybrid logical clock as a wall-clock timestamp, as opposed to recommending that applications use one of the user_data fields as an application-friendly timestamp.

(That said, I'm very open to using `timestamp` as an application timestamp -- user_data is a limited resource!)

Reviewer 2: TigerBeetle's cluster time HLC is definitely meant to be used as the authoritative wall-clock timestamp by the application.

It's good/safe, and also better than the API layer using different clocks in each stateless API instance, since TB can detect clock faults and clock skew and guarantee that a quorum of clocks feed into what is "true time" ("cluster time"), whereas stateless API instances might have unsynced clocks (or worse, might be partitioned asymmetrically from any clock sync protocol, with no way to detect that).

Another way of thinking about this, if TB's timestamps were not more trustworthy, then TB shouldn't be timestamping transactions, but instead each (unsynchronized) stateless API instance would be trusted to give its own view of time, and to timestamp transactions. But this is dangerous, exposing pending transfers to clock jumps if they happen in any instance in the API layer. And fixing this in the API layer would put more burden on what should be a simple stateless layer—designing the API layer to higher fault tolerances (i.e. implementing clock sync, as well as clock sync accuracy verification). Better for the smaller, centralized stateful component to take these burdens, than to place them on the many stateless components.

Finally, yet another way to see this, the timestamps here are used against recorded, historical, authoritative data from the financial system of record (TB, not stateless API layer). If the API layer instance's own clock seems to be behind/ahead, then that is simply the "clock fault model" playing out, and the API layer instance should realize it has strayed from the authority in the system.

Author: > Hmm, what if both timestamp_min and timestamp_max were exclusive instead of inclusive?
That way pagination still works as before.

The Query API will probably allow the user to choose the operator (e.g., `>` or `>=`), but for now, we need to make this decision.

I think inclusive ranges are more natural when dealing with time periods. Even with exclusive ranges, it would still require the user to set the correct field `timestamp_min` or `timestamp_max` for pagination depending on the sort order, so it's not a big win.

* Type is 128-bit unsigned integer (16 bytes)
* Must not be zero or `2^128 - 1`

### `timestamp_min`

The minimum [`Transfer.timestamp`](../transfers.md#timestamp) from which results will be returned, inclusive range.

> Optional, use zero for no lower bound filter by timestamp.

Reviewer: ```suggestion
Optional; set to zero to disable the lower-bound filter.
```
